File: src/workflows/workflowExecutionEngine.ts
/**
 * Converted from Python: core/workflows/workflow_execution_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function executeWorkflowPlan(plan: any, tick: any = 0): any {
  var executed: any[] = [];
  var index: any;
  var step: any;
  for ([index, step] of py.enumerate(py.slice(py.get(plan, "steps", []), null, 10000))) {
    py.listAppend(executed, {"step_id": py.toStr(py.get(step, "id", `step:${py.toStr(index)}`)), "action": py.toStr(py.get(step, "action", "")), "runtime": py.toStr(py.get(step, "runtime", "browser")), "completed": true, "tick": py.add(tick, index), "replay_index": index});
  }
  return {"objective": py.get(plan, "objective", ""), "executed": executed, "completed_count": py.len(executed), "bounded": true};
}

File: lib/src/workflows/workflow_execution_engine.dart
// Port of core/workflows/workflow_execution_engine.py

Map<String, dynamic> executeWorkflowPlan(
  Map<String, dynamic> plan, {
  int tick = 0,
}) {
  final List<Map<String, dynamic>> executed = <Map<String, dynamic>>[];

  final List<dynamic> steps = plan['steps'] is List
      ? List<dynamic>.from(plan['steps'] as List<dynamic>)
      : <dynamic>[];

  int index = 0;
  for (final dynamic stepRaw in steps.take(10000)) {
    final Map<String, dynamic> step = stepRaw is Map
        ? Map<String, dynamic>.from(stepRaw)
        : <String, dynamic>{};
    executed.add(<String, dynamic>{
      'step_id': '${step['id'] ?? 'step:$index'}',
      'action': '${step['action'] ?? ''}',
      'runtime': '${step['runtime'] ?? 'browser'}',
      'completed': true,
      'tick': tick + index,
      'replay_index': index,
    });
    index++;
  }

  return <String, dynamic>{
    'objective': plan['objective'] ?? '',
    'executed': executed,
    'completed_count': executed.length,
    'bounded': true,
  };
}
